libdnf/transaction/rpm_package.cpp

```diff
@@ -0,0 +1,199 @@
+            ti.reason as reason
+        FROM
+            trans_item ti
+        JOIN
+            trans t ON ti.trans_id = t.id
+        JOIN
+            rpm i USING (item_id)
+        WHERE
+            t.state = 1
+            AND ti.action not in (3, 5, 7, 10)
+            AND i.name = ?
+            AND i.arch = ?
+        ORDER BY
+            ti.trans_id DESC
+        LIMIT 1
+    )**";
+
+    if (arch != "") {
+        libdnf::utils::SQLite3::Query query(conn, sql);
+        query.bindv(name, arch);
+
+        if (query.step() == libdnf::utils::SQLite3::Statement::StepResult::ROW) {
+            auto action = static_cast<TransactionItemAction>(query.get<int64_t>("action"));
+            if (action == TransactionItemAction::REMOVE) {
+                return TransactionItemReason::UNKNOWN;
+            }
+            auto reason = static_cast<TransactionItemReason>(query.get<int64_t>("reason"));
+            return reason;
+        }
+    } else {
+        const char * arch_sql = R"**(
+            SELECT DISTINCT
+                arch
+            FROM
+                rpm
+            WHERE
+                name = ?
+        )**";
+
+        libdnf::utils::SQLite3::Query arch_query(conn, arch_sql);
+        arch_query.bindv(name);
+
+        TransactionItemReason result = TransactionItemReason::UNKNOWN;
+
+        while (arch_query.step() == libdnf::utils::SQLite3::Statement::StepResult::ROW) {
+            auto rpm_arch = arch_query.get<std::string>("arch");
+
+            libdnf::utils::SQLite3::Query query(conn, sql);
+            query.bindv(name, rpm_arch);
+            while (query.step() == libdnf::utils::SQLite3::Statement::StepResult::ROW) {
+                auto action = static_cast<TransactionItemAction>(query.get<int64_t>("action"));
+                if (action == TransactionItemAction::REMOVE) {
+                    continue;
+                }
+                auto reason = static_cast<TransactionItemReason>(query.get<int64_t>("reason"));
+                if (reason > result) {
+                    result = reason;
+                }
+            }
+        }
+        return result;
+    }
+    return TransactionItemReason::UNKNOWN;
+}
+*/
+
+/**
+ * Compare RPM packages
+ * This method doesn't care about compare package names
+ * \param other Package to compare with
+ * \return true if other package is newer (has higher version and/or epoch)
+ */
+bool Package::operator<(const Package & other) const {
+    // compare epochs
+    int32_t epoch_diff = other.get_epoch_int() - get_epoch_int();
+    if (epoch_diff > 0) {
+        return true;
+    } else if (epoch_diff < 0) {
+        return false;
+    }
+
+    // compare versions
+    std::stringstream versionThis(get_version());
+    std::stringstream versionOther(other.get_version());
+
+    std::string bufferThis;
+    std::string bufferOther;
+    while (std::getline(versionThis, bufferThis, '.') && std::getline(versionOther, bufferOther, '.')) {
```

Comment:
Why did you not use a `pool_evrcmp` functionality. This is like a reimplemented wheel with probably a different shape.

Reply:
This is an old code I did not touch except fixing variable / method names etc.
Not sure if pool_evrcmp() is usable, because the data is not in the pool.
On the other hand, it's a good find actually. I've added a TODO to replace the whole function with a different implementation, preferably an existing code.

Reply:
pool_evrcmp() requres pool to create tmp cache. Thanks for TODO